envoy/http/async_client.h

```diff
@@ -163,13 +170,52 @@ class AsyncClient {
      */
     virtual void reset() PURE;
 
+    /***
+     * Add destructor callback.
+     */
+    virtual void setDestructorCallback(StreamDestructorCallbacks callback) PURE;
+
+    /***
+     * Remove previously set destructor callback.
+     */
+    virtual void removeDestructorCallback() PURE;
+
+    /***
+     * Add watermark callbacks.
+     */
+    virtual void setWatermarkCallbacks(DecoderFilterWatermarkCallbacks& callbacks) PURE;
+
+    /***
+     * Remove previously set watermark callbacks.
+     */
+    virtual void removeWatermarkCallbacks() PURE;
+
     /***
      * @returns if the stream has enough buffered outbound data to be over the configured buffer
      * limits
      */
     virtual bool isAboveWriteBufferHighWatermark() const PURE;
   };
 
+  /***
+   * An interface for ongoing asynchronous requests.
+   */
+  class OngoingRequest : public virtual Request, public virtual Stream {
+  public:
+    /***
+     * Take ownership of data, and sends it to the underlying stream.
+     * @param data owned buffer to pass to upstream.
+     * @param end_stream whether to end the stream.
+     */
+    virtual void captureAndSendData(Buffer::InstancePtr&& data, bool end_stream) PURE;
```

Comment:
There are other ways to consume data from a buffer, e.g. taking the underlying slices which might be a better approach.

Reply:
Could you clarify? We need to copy the original buffer into each of the shadows, and ownership needs to be retained by the new request.